File: services/tileMath.ts
import { GeoLocation, TileCoordinate } from '../types';

export const latLonToTile = (lat: number, lon: number, zoom: number): TileCoordinate => {
  const n = Math.pow(2, zoom);
  const x = Math.floor(n * ((lon + 180) / 360));
  const latRad = (lat * Math.PI) / 180;
  const y = Math.floor(
    (n * (1 - Math.log(Math.tan(latRad) + 1 / Math.cos(latRad)) / Math.PI)) / 2
  );
  return { x, y, z: zoom };
};

// Returns floating point tile coordinates for smooth panning math
export const latLonToTilePoint = (lat: number, lon: number, zoom: number): { x: number; y: number } => {
  const n = Math.pow(2, zoom);
  const x = n * ((lon + 180) / 360);
  const latRad = (lat * Math.PI) / 180;
  const y = (n * (1 - Math.log(Math.tan(latRad) + 1 / Math.cos(latRad)) / Math.PI)) / 2;
  return { x, y };
};

export const tileToLatLon = (x: number, y: number, z: number): GeoLocation => {
  const n = Math.pow(2, z);
  const lon = (x / n) * 360 - 180;
  const latRad = Math.atan(Math.sinh(Math.PI * (1 - (2 * y) / n)));
  const lat = (latRad * 180) / Math.PI;
  return { lat, lon };
};

// Calculate how many meters per pixel at a given latitude and zoom
// (Approximate for spherical mercator)
export const metersPerPixel = (lat: number, zoom: number): number => {
  return (156543.03392 * Math.cos((lat * Math.PI) / 180)) / Math.pow(2, zoom);
};

export const getTileUrl = (template: string, x: number, y: number, z: number): string => {
  return template
    .replace('{z}', z.toString())
    .replace('{x}', x.toString())
    .replace('{y}', y.toString());
};

File: android-app/app/src/main/java/com/orbitwall/wallpaper/TileMath.kt
package com.orbitwall.wallpaper

import com.orbitwall.model.GeoLocation

object TileMath {
    fun latLonToTilePoint(lat: Double, lon: Double, zoom: Int): Pair<Double, Double> {
        val n = Math.pow(2.0, zoom.toDouble())
        val x = n * ((lon + 180.0) / 360.0)
        val latRad = Math.toRadians(lat)
        val y = n * (1 - (Math.log(Math.tan(latRad) + 1 / Math.cos(latRad)) / Math.PI)) / 2.0
        return x to y
    }

    fun tileToLatLon(x: Double, y: Double, zoom: Int): GeoLocation {
        val n = Math.pow(2.0, zoom.toDouble())
        val lon = x / n * 360.0 - 180.0
        val latRad = Math.atan(Math.sinh(Math.PI * (1 - 2 * y / n)))
        val lat = Math.toDegrees(latRad)
        return GeoLocation(lat = lat, lon = lon)
    }

    fun metersPerPixel(lat: Double, zoom: Int): Double {
        return (156543.03392 * Math.cos(Math.toRadians(lat))) / Math.pow(2.0, zoom.toDouble())
    }
}
